File: Graphs/733_Flood_Fill/solution733.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int start_color = image[sr][sc];
        
        if (start_color == color) return image;
        
        flood_fill(image, sr, sc, start_color, color);
        
        return image;
    }
    
private:
    void flood_fill(vector<vector<int>>& image, int x, int y, int start_color, int color) {
        if (x < 0 || x >= image.size() || y < 0 || y >= image[0].size()) return;
        
        if (image[x][y] != start_color) return;
        
        image[x][y] = color;
        
        flood_fill(image, x - 1, y, start_color, color);
        flood_fill(image, x + 1, y, start_color, color);
        flood_fill(image, x, y + 1, start_color, color);
        flood_fill(image, x, y - 1, start_color, color);
    }
};

// Time Complexity:
// Space Complexity:

// Runtime1: 4ms, beats 84.72%
// Memory1: 16.85MB, beats 86.60%

// Runtime2: 0ms, beats 100%
// Memory2: 16.72MB, beats 94.63%

// Runtime3: 0ms, beats 100%
// Memory3: 16.78MB, beats 94.63%

File: Graphs/733_Flood_Fill/solution733.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        
        start_color = image[sr][sc]
        
        def flood_fill(x, y):
            if x < 0 or x >= len(image): return
            if y < 0 or y >= len(image[0]): return
            
            if image[x][y] == color: return
            if image[x][y] != start_color: return
            
            image[x][y] = color
            
            flood_fill(x-1, y)
            flood_fill(x+1, y)
            flood_fill(x, y+1)
            flood_fill(x, y-1)
        
        flood_fill(sr, sc)
        return image

# Runtime1: 64ms, beats 74.66%
# Memory1: 17MB, beats 5.87%

# Runtime2: 64ms, beats 74.80%
# Memory2: 16.92MB, beats 6.29%

# Runtime3: 62ms, beats 83.67%
# Memory3: 16.86MB, beats 22.46%